src/Features/CSharp/Portable/ExtractLocalMethod/CSharpExtractLocalMethodCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,83 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Composition;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.CodeRefactorings.ExtractMethod;
+using Microsoft.CodeAnalysis.CSharp.CodeStyle;
+using Microsoft.CodeAnalysis.ExtractMethod;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.ExtractLocalMethod
+{
+    [ExportCodeRefactoringProvider(LanguageNames.CSharp,
+        Name = PredefinedCodeRefactoringProviderNames.ExtractLocalMethod), Shared]
+    internal class CSharpExtractLocalMethodCodeRefactoringProvider : ExtractMethodCodeRefactoringProvider
+    {
+        [ImportingConstructor]
+        public CSharpExtractLocalMethodCodeRefactoringProvider()
+        {
+        }
+
+        protected override async Task<(CodeAction action, string methodBlock)> GetCodeActionAsync(
+            Document document,
+            TextSpan textSpan,
+            CancellationToken cancellationToken)
+        {
+            var documentOptions = await document.GetOptionsAsync(cancellationToken).ConfigureAwait(false);
+            var preferStatic = documentOptions.GetOption(CSharpCodeStyleOptions.PreferStaticLocalFunction).Value;
+
+            var result = await ExtractMethodService.ExtractMethodAsync(
+                document,
+                textSpan,
+                extractLocalMethod: true,
+                cancellationToken: cancellationToken).ConfigureAwait(false);
+            Contract.ThrowIfNull(result);
+
+            if (result.Succeeded || result.SucceededWithSuggestion)
+            {
+                // We insert an empty line between the generated local method and the previous statements if there is not one already.
+                var (resultDocument, resultMethodDeclarationNode, resultInvocationNameToken) = await InsertNewLineBeforeLocalMethodIfNecessaryAsync(result, cancellationToken).ConfigureAwait(false);
```

Comment:
not sure why this is done here.  feels like it should be done in the callback that is passed to MyCodeAction.